### DISCUSSION TITLE
Native C calling Java method

### DISCUSSION THREAD

**Discussion by paulocabrita-ionos** [jnr/jnr-ffi#320] [General]
Hello.

I am having some issues in transforming a Pointer into a java object through a callback function (C is calling a Java Method).

My flow is as follows:
// calling native function. xpto is the native library. The second parameter is the Java method that the C native will call.
int ret = xpto.setAuthenticate(this.context, new JavaAuthenticateImpl());

// My class with the method called by Native C. I know that C calls this method because I can get the pointer address in Java. How to build a java object from the Pointer??

public class JavaAuthenticateImpl implements JavaAuthenticateCb {

    @Override
    public boolean javaAuthenticateCb(Pointer context, Pointer token) {

        // I know the token is a "MyToken" type but how to convert it??

        System.out.println("Authenticate: " + token);
        return false;
    }
}

// My interface.
public interface JavaAuthenticateCb {
    @Delegate
    public boolean javaAuthenticateCb(Pointer context, Pointer token);
}

Are they some examples? I just can't find them.

Another thing, my native library has an include file with the description of all the C structs and they are ABI compliant. But, from my knowledge, JNR does not have the option to use ABI. But maybe I'm wrong.

How can I build a java object from a pointer inside a Java method that is called by a native function?

Thanks.

**Comment by headius** [comment]
I'll convert this to a discussion, and we can go from there.

**Comment by headius** [comment]
There's a small example of using an interface to callback from C here: https://github.com/jnr/jnr-ffi-examples/blob/master/qsort/src/main/java/qsort/Qsort.java

Does that help you?

**Reply by paulocabrita-ionos** [reply]
Thanks @headius . I already did take a look at the link and it helped.

Like I said previously, I am using an include file (.h) defined with ABI. So it was fair straight forward to do the conversion type. :)

We can close this.

I have another issue though. But I will open another discussion.

Many thanks.
